Tasks/DeployAzureResourceGroup/MachineGroupAgentExtensionManager.ts

```diff
@@ -0,0 +1,179 @@
+                this.log(tl.loc("AddExtension", extensionParameters["extensionName"], vmName));
+                var extensionName = extensionParameters["extensionName"];
+                var parameters = extensionParameters["parameters"];
+                var callback = this.givePostOperationCallBack(extensionName, vmName, operation);
+                this.computeClient.virtualMachineExtensions.createOrUpdate(resourceGroupName, vmName, extensionName, parameters, callback);
+            }
+            return this.deferred.promise;
+        }
+        catch (exception) {
+            this.log(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            tl.setResult(tl.TaskResult.Failed, tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            this.deferred.resolve(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+        }
+    }
+
+    public async deleteMGExtension() {
+        try {
+            var operation = "uninstallation";
+            this.deferred = Q.defer<string>();
+            var listOfVms = await this.azureUtils.getVMDetails();
+            this.vmCount = listOfVms.length;
+            if (this.vmCount == 0) {
+                this.deferred.resolve("");
+            }
+            for (var i = 0; i < listOfVms.length; i++) {
+                var vmName = listOfVms[i]["name"];
+                var resourceGroupName = this.taskParameters.resourceGroupName;
+                var extensionParameters = this.formExtensionParameters(listOfVms[i], operation);
+                this.log(tl.loc("DeleteExtension", extensionParameters["extensionName"], vmName));
+                var extensionName = extensionParameters["extensionName"];
+                var callback = this.givePostOperationCallBack(extensionName, vmName, operation);
+                this.computeClient.virtualMachineExtensions.deleteMethod(resourceGroupName, vmName, extensionName, callback);
+            }
+            return this.deferred.promise;
+        }
+        catch (exception) {
+            this.log(tl.loc("MGAgentOperationOnAllVMsFailed", operation, exception.message));
+            tl.setResult(tl.TaskResult.Failed, tl.loc("MGAgentOperationOnAllVMsFailed", operation, exception.message));
+            this.deferred.reject(tl.loc("MGAgentOperationOnAllVMsFailed", operation, exception.message));
+        }
+    }
+
+    private setTaskResult(operation) {
+        if (this.failureCount + this.successCount == this.vmCount) {
+            if (this.failureCount > 0) {
+                this.log(tl.loc("MGAgentOperationOnAllVMsFailed", operation, ""));
+                this.deferred.reject(operation);
+            }
+            else {
+                this.log(tl.loc("MGAgentOperationOnAllVMsSucceeded", operation));
+                this.deferred.resolve(operation);
+            }
+        }
+    }
+
+    private givePostOperationCallBack(extensionName, vmName, operation) {
```

Comment:
Since 'createOrUpdate' is a task action and I have added methods that begin with that name, I have changed the name to 'PostOperationCallBack' instead